pilot/pkg/networking/core/v1alpha3/tracing.go

```diff
@@ -0,0 +1,519 @@
+		return buildHCMTracing(pushCtx, providerCfg.Name, provider.Datadog.Service, provider.Datadog.Port, provider.Datadog.MaxTagLength, datadogConfigGen)
+	case *meshconfig.MeshConfig_ExtensionProvider_Lightstep:
+		return buildHCMTracing(pushCtx, providerCfg.Name, provider.Lightstep.Service, provider.Lightstep.Port, provider.Lightstep.MaxTagLength,
+			func(clusterName string) (*anypb.Any, error) {
+				lc := &tracingcfg.LightstepConfig{
+					CollectorCluster: clusterName,
+					AccessTokenFile:  provider.Lightstep.AccessToken,
+				}
+				return anypb.New(lc)
+			})
+
+	case *meshconfig.MeshConfig_ExtensionProvider_Opencensus:
+		return buildHCMTracingOpenCensus(providerCfg.Name, provider.Opencensus.MaxTagLength, func() (*anypb.Any, error) {
+			oc := &tracingcfg.OpenCensusConfig{
+				OcagentAddress:         fmt.Sprintf("%s:%d", provider.Opencensus.Service, provider.Opencensus.Port),
+				OcagentExporterEnabled: true,
+				IncomingTraceContext:   convert(provider.Opencensus.Context),
+				OutgoingTraceContext:   convert(provider.Opencensus.Context),
+			}
+
+			return anypb.New(oc)
+		})
+
+	case *meshconfig.MeshConfig_ExtensionProvider_Stackdriver:
+		return buildHCMTracingOpenCensus(providerCfg.Name, provider.Stackdriver.MaxTagLength, func() (*anypb.Any, error) {
+			proj, ok := meta.PlatformMetadata[platform.GCPProject]
+			if !ok {
+				proj, ok = meta.PlatformMetadata[platform.GCPProjectNumber]
+			}
+			if !ok {
+				return nil, fmt.Errorf("could not configure Stackdriver tracer - unknown project id")
+			}
+
+			sd := &tracingcfg.OpenCensusConfig{
+				StackdriverExporterEnabled: true,
+				StackdriverProjectId:       proj,
+				IncomingTraceContext:       allContexts,
+				OutgoingTraceContext:       allContexts,
+				StdoutExporterEnabled:      provider.Stackdriver.Debug,
+				TraceConfig: &opb.TraceConfig{
+					MaxNumberOfAnnotations:   200,
+					MaxNumberOfAttributes:    200,
+					MaxNumberOfMessageEvents: 200,
+				},
+			}
+
+			if meta.StsPort != "" {
+				stsPort, err := strconv.Atoi(meta.StsPort)
+				if err != nil || stsPort < 1 {
+					return nil, fmt.Errorf("could not configure Stackdriver tracer - bad sts port: %v", err)
+				}
+				sd.StackdriverGrpcService = &envoy_config_core_v3.GrpcService{
+					InitialMetadata: []*envoy_config_core_v3.HeaderValue{
+						{
+							Key:   "x-goog-user-project",
+							Value: proj,
+						},
+					},
+					TargetSpecifier: &envoy_config_core_v3.GrpcService_GoogleGrpc_{
+						GoogleGrpc: &envoy_config_core_v3.GrpcService_GoogleGrpc{
+							TargetUri:  "cloudtrace.googleapis.com",
+							StatPrefix: "oc_stackdriver_tracer",
+							ChannelCredentials: &envoy_config_core_v3.GrpcService_GoogleGrpc_ChannelCredentials{
+								CredentialSpecifier: &envoy_config_core_v3.GrpcService_GoogleGrpc_ChannelCredentials_SslCredentials{
+									SslCredentials: &envoy_config_core_v3.GrpcService_GoogleGrpc_SslCredentials{
+										RootCerts: &envoy_config_core_v3.DataSource{
```

Comment:
Nit we don't need ot set the root certs here. Just specifying an empty ssl credentials is good enough. 
```
SslCredentials: &envoy_config_core_v3.GrpcService_GoogleGrpc_SslCredentials{}
```

We need to clean the bootstrap as well.

Reply:
cleaned it up here.